### DISCUSSION TITLE
More flexibility in ACSFs generation

### DISCUSSION THREAD

**Discussion by soniasalomoni** [MarsalekGroup/aml#15] [General]
Hi, 
I'm thinking of implementing n2p2's polynomial ACSFs in aml, and I was wondering if you were already developing some kind of features to include them and also if you were planning to give the users flexibility in selecting ACSFs since the ACSFs generation is hardcoded in aml's current version.
Best,
Sonia

**Comment by OndrejMarsalek** [comment]
Dear Sonia,

Thank you for your interest in the code. In the example that is available in this released version of the code, only the default set, as used in the PNAS paper, is used. Although it is not clear from the example itself, it is already possible to customize the ACSFs that you use. There are two options:
1) In the template of `input.nn`, you do not include the `{acsf:}` formatting field and just put in valid n2p2 input defining the symmetry functions that you want. That means that AML does not help you with the generation of the ACSFs, but you have full flexibility.
2) With a small modification, you can also use AML to generate a different set for you. You leave the `{acsf:}` formatting field in the template, and modify the code around here:
https://github.com/MarsalekGroup/aml/blob/main/aml/mlp.py#L383
where you can provide different `radials` and `angulars`, as generated by these functions:
https://github.com/MarsalekGroup/aml/blob/main/aml/acsf.py#L43
https://github.com/MarsalekGroup/aml/blob/main/aml/acsf.py#L57
This is not ideal but does give you some flexibility in principle.

In future versions, we plan an easier way to pass custom ACSFs to the model as well as support for generating more sets of ACSFs. wACSFs are already implemented and polynomial ACSFs will be easy to add. If you would like to do that locally, it should not be very difficult based on what is already in the `acsf` module. For integration into AML proper, we might want to first look at the structure of the module, perhaps things could be done a bit differently with several types of ACSFs.

All the best,
Ondrej

**Reply by cschran** [reply]
Hi both,

as a quick addition to Ondrej's comment, I would like to add that cp2k does not support wACSFs or polynomial ACSFs at the moment. You would need to use lammps+n2p2 for both these combinations in order to run simulations, however with a single model as n2p2 does not have committees yet.

Best,

Christoph

**Reply by OndrejMarsalek** [reply]
Thanks for the clarification, Chris, I should have mentioned that.

This is important from the practical point of view, of course. It means you can train a committee of models based on wACSFs or polynomial ACSFs using AML, but can't use it for MD at the moment. If you are willing and able to add enough data to your training set and don't mind losing the benefits of the committee, you could in principle use just one of the models, perhaps after some additional validation, in current LAMMPS/n2p2. Eventually, committees should make their way to n2p2 and even the LAMMPS interface.
